File: compiler/bench/heapify.fs
\ expected: 352
\ Build max-heap from 100K elements, return max

100000 constant N
create heap N cells allot

: swap-heap ( i j -- )
  cells heap + swap cells heap +
  over @ over @
  rot ! swap ! ;

: sift-down ( i -- )
  begin
    dup 2* 1+ dup N < while
    dup 1+ N < if
      dup cells heap + @
      over 1+ cells heap + @
      < if 1+ then
    then
    2dup cells heap + @
    swap cells heap + @ < if
      2dup swap-heap
      nip
    else
      2drop exit
    then
  repeat
  2drop ;

: heapify ( -- )
  N 2/ 1- 0 do
    N 2/ 1- i - sift-down
  loop ;

: init-heap ( -- )
  N 0 do
    i 17 mod i 23 mod * N mod i cells heap + !
  loop ;

: main init-heap heapify heap @ . cr ;

File: compiler/bench/heapify.c
// expected: 352
#include <stdio.h>
#define N 100000

long heap[N];

void swap_heap(int i, int j) {
    long t = heap[i];
    heap[i] = heap[j];
    heap[j] = t;
}

void sift_down(int i) {
    while (2 * i + 1 < N) {
        int j = 2 * i + 1;
        if (j + 1 < N && heap[j] < heap[j + 1]) j++;
        if (heap[i] < heap[j]) {
            swap_heap(i, j);
            i = j;
        } else {
            break;
        }
    }
}

void heapify() {
    for (int i = N / 2 - 1; i >= 0; i--) {
        sift_down(i);
    }
}

void init_heap() {
    for (int i = 0; i < N; i++) {
        heap[i] = ((long)(i % 17) * (i % 23)) % N;
    }
}

int main() {
    init_heap();
    heapify();
    printf("%ld\n", heap[0]);
    return 0;
}
